### PR TITLE
ARROW-12706: [Python] Drop Python 3.6 support

### PR DESCRIPTION
**Pull request by pitrou**

### CODE REVIEW

**Comment by pitrou** [comment]
@github-actions crossbow submit *py*

**Comment by pitrou** [comment]
@github-actions crossbow submit -g wheel

**Comment by pitrou** [comment]
@github-actions crossbow submit *py*

**Comment by pitrou** [comment]
@github-actions crossbow submit -g wheel

**Comment by pitrou** [comment]
@github-actions crossbow submit test-ubuntu-20.04-python-3

**Comment by pitrou** [comment]
@github-actions crossbow submit -g wheel

File: python/pyarrow/compat.pxi [hunk 1]
```diff
@@ -31,12 +29,8 @@ def encode_file_path(path):
...
 
 
-if sys.version_info >= (3, 7):
-    # Starting with Python 3.7, dicts are guaranteed to be insertion-ordered.
-    ordered_dict = dict
-else:
-    import collections
-    ordered_dict = collections.OrderedDict
+# Starting with Python 3.7, dicts are guaranteed to be insertion-ordered.
+ordered_dict = dict
```
Comment (jorisvandenbossche) [inline python/pyarrow/compat.pxi:33]: Not necessarily here, but I suppose we can also remove `ordered_dict` alltogether, and use `dict(..)` instead in the few places it is being used?

**Review by jorisvandenbossche — APPROVED** [review]
Looks good!

File: python/pyarrow/compat.pxi (hunk 1 above)
Comment (pitrou) [reply python/pyarrow/compat.pxi:33]: `ordered_dict` has a richer API than `dict`, so theoretically we cannot return `dict` instead without breaking compatibility. Not sure that matters in practice, though.

Comment (jorisvandenbossche) [reply python/pyarrow/compat.pxi:33]: > `ordered_dict` has a richer API than `dict`, so theoretically we cannot return `dict` instead without breaking compatibility. Not sure that matters in practice, though.

Ah, yes indeed. It's currently being used in `Table/RecordBatch.to_pydict` and `KeyValueMetadata.to_dict`. 
Personally, I would maybe just break this at some point

**Pull request closed without merging**